optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolver.java

```diff
@@ -149,16 +149,20 @@ public final void solve(Solution planningProblem) {
             throw new IllegalArgumentException("The planningProblem (" + planningProblem
                     + ") must not be null.");
         }
-        solverScope.setBestSolution(planningProblem);
-        outerSolvingStarted(solverScope);
-        boolean restartSolver = true;
-        while (restartSolver) {
-            solvingStarted(solverScope);
-            runPhases();
-            solvingEnded(solverScope);
-            restartSolver = checkProblemFactChanges();
+        try {
+            solverScope.setBestSolution(planningProblem);
+            outerSolvingStarted(solverScope);
+            boolean restartSolver = true;
+            while (restartSolver) {
+                solvingStarted(solverScope);
+                runPhases();
+                solvingEnded(solverScope);
+                restartSolver = checkProblemFactChanges();
+            }
+            outerSolvingEnded(solverScope);
+        } finally {
+            solving.set(false);
```

Comment:
@ge0ffrey You are right - it's an user exception thrown in a custom implementation of IncrementalScoreCalculator.

Actually there is one thread in which planning is executed. The planning is limited to, say 5 minutes, so we rely on Solver.isSolving() and use Solver.iterminateEarly() from other thread to get then the best solution and reuse the Solver for the next task.

In our case the Solver is kind of a service which instantiated and managed by Spring framework - so I particularly did not consider the possibility of restarting of the Solver - it seems be bulky work. And I also thought it be useful in common case - to guaranteed change the state of Solver if something goes wrong.

Anyway I'll ask more experienced coworker about restarting Solver, thanks for advise.